source/arch/encoding_simd_sse41.c

```diff
@@ -0,0 +1,396 @@
+        14, 13, 12,
+
+        0xFF, 0xFF, 0xFF, 0xFF, /* Zero out the top 4 bytes of the lane */
+        2,  1,  0,
+        6,  5,  4,
+        10,  9,  8,
+        14, 13, 12
+        /* LSB */
+        /* clang-format on */
+    }};
+    __m256i shufvec = _mm256_load_si256((__m256i const *)&shufvec_buf);
+
+    dwords = _mm256_shuffle_epi8(dwords, shufvec);
+    /*
+     * Now shuffle the 32-bit words:
+     * A B C 0 D E F 0 -> 0 0 A B C D E F
+     */
+    __m256i shuf32 = _mm256_set_epi32(0, 0, 7, 6, 5, 3, 2, 1);
+
+    dwords = _mm256_permutevar8x32_epi32(dwords, shuf32);
+
+    return dwords;
+}
+
+static inline bool decode(const unsigned char *in, unsigned char *out) {
+    __m256i vec = _mm256_loadu_si256((__m256i const *)in);
+    if (!decode_vec(&vec)) {
+        return false;
+    }
+    vec = pack_vec(vec);
+
+    /*
+     * We'll do overlapping writes to get both the low 128 bits and the high 64-bits written.
+     * Input (memory order): 0 1 2 3 4 5 - - (dwords)
+     * Input (little endian) - - 5 4 3 2 1 0
+     * Output in memory:
+     * [0 1 2 3] [4 5]
+     */
+    __m128i lo = _mm256_extracti128_si256(vec, 0);
+    uint64_t hi = _mm256_extract_epi64(vec, 2);
+
+    _mm_storeu_si128((__m128i *)out, lo);
+    memcpy(out + 16, &hi, sizeof(hi));
+
+    return true;
+}
+
+size_t aws_common_private_base64_decode_sse41(const unsigned char *in, unsigned char *out, size_t len) {
+    if (len % 4)
+        return -1;
+
+    size_t outlen = 0;
+    while (len > 32) {
+        if (!decode(in, out))
+            return -1;
+        len -= 32;
+        in += 32;
+        out += 24;
+        outlen += 24;
+    }
+
+    if (len > 0) {
+        unsigned char tmp_in[32];
+        unsigned char tmp_out[24];
+
+        memset(tmp_out, 0xEE, sizeof(tmp_out));
+
+        /* We need to ensure the vector contains valid b64 characters */
+        memset(tmp_in, 'A', sizeof(tmp_in));
+        memcpy(tmp_in, in, len);
+
+        size_t final_out = (3 * len) / 4;
+
+        /* Check for end-of-string padding (up to 2 characters) */
+        for (int i = 0; i < 2; i++) {
+            if (tmp_in[len - 1] == '=') {
+                tmp_in[len - 1] = 'A'; /* make sure the inner loop doesn't bail out */
+                len--;
+                final_out--;
+            }
+        }
+
+        if (!decode(tmp_in, tmp_out))
+            return -1;
+
+        /* Check that there are no trailing ones bits */
+        for (size_t i = final_out; i < sizeof(tmp_out); i++) {
+            if (tmp_out[i])
+                return -1;
+        }
+
+        memcpy(out, tmp_out, final_out);
+        outlen += final_out;
+    }
+    return outlen;
+}
+
+/***** Encode logic *****/
+static inline __m256i encode_chars(__m256i in) {
+    __m256i tmp;
+
+    tmp = translate_range(in, 0, 25, 'A');
+    tmp |= translate_range(in, 26, 26 + 25, 'a');
+    tmp |= translate_range(in, 52, 61, '0');
+    tmp |= translate_exact(in, 62, '+');
+    tmp |= translate_exact(in, 63, '/');
```

Comment:
As above: can magic numbers please be avoided?

Reply:
Again, this is an encoding table, where else would the magic numbers go?

Reply:
Add comments. Either you made them up, or you got them from somewhere. If the former, say so. If the latter, add a link to your source.